src/encoded/static/components/cart/cart.js

```diff
@@ -979,256 +888,646 @@ PagerArea.propTypes = {
+ *
+ * @return {object} Facet terms keyed by displayed facet fields.
+ */
+const generateFacetTermsTemplate = () => {
+    const termTemplate = {};
+    displayedFacetFields.forEach((facetField) => {
+        termTemplate[facetField.field] = [];
+    });
+    return termTemplate;
+};
+
+
+/**
+ * Retrieve partial file objects for all given datasets, as well as a list of datasets viewable at
+ * the user's access level -- needed for shared carts.
+ * @param {array} datasetsIds Array of dataset @ids to retrieve
+ * @param {func} facetLoadHandler Called with progress in loading chunks of datasets1
+ * @param {func} fetch System fetch
+ * @param {object} session System session
+ *
+ * @return {promise}:
+ * {
+ *      datasetFiles - Array of all parital file objects in all datasets
+ *      datasets - Array of all datasets viewable at user's access level; subset of `datasetsIds`
+ * }
+ */
+const retrieveDatasetsFiles = (datasetsIds, facetProgressHandler, fetch, session) => {
+    // Break incoming array of dataset @ids into manageable chunks of arrays, each with
+    // CHUNK_SIZE elements. Search results from all these chunks get consolidated before returning
+    // a promise.
+    const CHUNK_SIZE = 500;
+    const chunks = [];
+    for (let datasetIndex = 0; datasetIndex < datasetsIds.length; datasetIndex += CHUNK_SIZE) {
+        chunks.push(datasetsIds.slice(datasetIndex, datasetIndex + CHUNK_SIZE));
+    }
+
+    // Using the arrays of dataset @id arrays, do a sequence of searches of CHUNK_SIZE datasets
+    // adding to extract information to display the facets, search results, and visualization.
+    facetProgressHandler(null);
+    return chunks.reduce((promiseChain, currentChunk, currentChunkIndex) => (
+        // As each experiment search-result promise resolves, add its results to the array of
+        // partial files and facets in `accumulatingResults`.
+        promiseChain.then(accumulatingResults => (
+            // Request one chunk of datasets. `currentResults` contains the request search
+            // results including the partial file objects we need.
+            requestDatasets(currentChunk, fetch, session).then((currentResults) => {
+                // Update progress on each chunk.
+                facetProgressHandler(Math.round(((currentChunkIndex + 1) / chunks.length) * 100));
+
+                // Add the chunk's worth of results to the array of partial files and facets
+                // we're accumulating.
+                return {
+                    datasetFiles: addToAccumulatingFiles(accumulatingResults.datasetFiles, currentResults),
+                    datasets: accumulatingResults.datasets.concat(currentResults['@graph'].map(experiment => experiment['@id'])),
+                };
+            })
+        ))
+    ), Promise.resolve({ datasetFiles: [], datasets: [] })).then(({ datasetFiles, datasets }) => {
+        facetProgressHandler(-1);
+        return { datasetFiles, datasets };
+    });
+};
```

Comment:
I think this isn't useful, but `requestFiles` and `requestObjects` in objectutils.js do a similar operation: https://github.com/ENCODE-DCC/encoded/blob/dev/src/encoded/static/components/objectutils.js#L209. It doesn't handle the progress bar the way you might need, and it also handles filtering and duplicate @ids which seem not relevant here. But, it is chunking data and fetching it!

Reply:
They are similar and I’m sure could be combined. But this needed just that little bit more and I didn’t want to touch functions used elsewhere yet.